contracts/src/L1/rollup/ScrollChain.sol

```diff
@@ -441,56 +568,181 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
+        internal
+        view
+        returns (
+            uint256 batchPtr,
+            bytes32 _batchHash,
+            uint256 _batchIndex,
+            uint256 _totalL1MessagesPoppedOverall
+        )
+    {
+        // load version from batch header, it is always the first byte.
+        uint256 version;
+        assembly {
+            version := shr(248, calldataload(_batchHeader.offset))
+        }
 
-        // compute batch hash
-        _batchHash = BatchHeaderV0Codec.computeBatchHash(memPtr, _length);
+        // version should be always 0 or 1 in current code
+        uint256 _length;
+        if (version == 0) {
+            (batchPtr, _length) = BatchHeaderV0Codec.loadAndValidate(_batchHeader);
+            _batchHash = BatchHeaderV0Codec.computeBatchHash(batchPtr, _length);
+            _batchIndex = BatchHeaderV0Codec.batchIndex(batchPtr);
+        } else if (version == 1) {
+            (batchPtr, _length) = BatchHeaderV1Codec.loadAndValidate(_batchHeader);
+            _batchHash = BatchHeaderV1Codec.computeBatchHash(batchPtr, _length);
+            _batchIndex = BatchHeaderV1Codec.batchIndex(batchPtr);
+        } else {
+            revert ErrorInvalidBatchHeaderVersion();
+        }
+        // only check when genesis is imported
+        if (committedBatches[_batchIndex] != _batchHash && finalizedStateRoots[0] != bytes32(0)) {
+            revert ErrorIncorrectBatchHash();
+        }
+        _totalL1MessagesPoppedOverall = BatchHeaderV0Codec.totalL1MessagePopped(batchPtr);
     }
 
-    /// @dev Internal function to commit a chunk.
-    /// @param memPtr The start memory offset to store list of `dataHash`.
+    /// @dev Internal function to commit a chunk with version 0.
     /// @param _chunk The encoded chunk to commit.
     /// @param _totalL1MessagesPoppedInBatch The total number of L1 messages popped in current batch.
     /// @param _totalL1MessagesPoppedOverall The total number of L1 messages popped in all batches including current batch.
     /// @param _skippedL1MessageBitmap The bitmap indicates whether each L1 message is skipped or not.
+    /// @return _dataHash The computed data hash for this chunk.
     /// @return _totalNumL1MessagesInChunk The total number of L1 message popped in current chunk
-    function _commitChunk(
-        uint256 memPtr,
+    function _commitChunkV0(
         bytes memory _chunk,
         uint256 _totalL1MessagesPoppedInBatch,
         uint256 _totalL1MessagesPoppedOverall,
         bytes calldata _skippedL1MessageBitmap
-    ) internal view returns (uint256 _totalNumL1MessagesInChunk) {
+    ) internal view returns (bytes32 _dataHash, uint256 _totalNumL1MessagesInChunk) {
         uint256 chunkPtr;
         uint256 startDataPtr;
         uint256 dataPtr;
-        uint256 blockPtr;
 
         assembly {
             dataPtr := mload(0x40)
             startDataPtr := dataPtr
             chunkPtr := add(_chunk, 0x20) // skip chunkLength
```

Comment:
Is this the same as `_chunk.offset`?

Reply:
`_chunk` is `memory`, no `_chunk.offset` exists.